apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+        words(CompressedTF)
+    ),
+    KeyFamily = [{0, <<>>} | Indices],
+    {ok, Batch} = rocksdb:batch(),
+    try
+        Ndeleted = lists:foldl(
+            fun({MsgKey, Val}, Acc) ->
+                case Selector(Val) of
+                    true ->
+                        do_delete(CF, Batch, Static, KeyFamily, MsgKey),
+                        Acc + 1;
+                    false ->
+                        Acc
+                end
+            end,
+            0,
+            KVs
+        ),
+        case rocksdb:write_batch(DB, Batch, []) of
+            ok ->
+                {ok, It, Ndeleted, length(KVs)};
+            {error, {error, Reason}} ->
+                {error, unrecoverable, {rocksdb, Reason}}
+        end
+    after
+        rocksdb:release_batch(Batch)
+    end.
+
+do_delete(CF, Batch, Static, KeyFamily, MsgKey) ->
+    TS = match_ds_key(Static, MsgKey),
+    lists:foreach(
+        fun({WildcardIdx, Hash}) ->
+            ok = rocksdb:batch_delete(Batch, CF, mk_key(Static, WildcardIdx, Hash, TS))
+        end,
+        KeyFamily
+    ).
+
+%%%%%%%% Iteration %%%%%%%%%%
+
+init_iterators(S, #it{static_index = Static, compressed_tf = CompressedTF}) ->
+    do_init_iterators(S, Static, words(CompressedTF), 1).
+
+do_init_iterators(S, Static, ['+' | TopicFilter], WildcardLevel) ->
+    %% Ignore wildcard levels in the topic filter:
+    do_init_iterators(S, Static, TopicFilter, WildcardLevel + 1);
+do_init_iterators(S, Static, [Constraint | TopicFilter], WildcardLevel) ->
+    %% Create iterator for the index stream:
+    #s{hash_bytes = HashBytes, db = DB, data_cf = DataCF} = S,
+    Hash = hash(HashBytes, Constraint),
+    {ok, ItHandle} = rocksdb:iterator(DB, DataCF, get_key_range(Static, WildcardLevel, Hash)),
+    It = #l{
+        n = WildcardLevel,
+        handle = ItHandle,
+        hash = Hash
+    },
+    [It | do_init_iterators(S, Static, TopicFilter, WildcardLevel + 1)];
+do_init_iterators(S, Static, [], _WildcardLevel) ->
+    %% Create an iterator for the data stream:
+    #s{db = DB, data_cf = DataCF} = S,
+    Hash = <<>>,
+    {ok, ItHandle} = rocksdb:iterator(DB, DataCF, get_key_range(Static, 0, Hash)),
+    [
+        #l{
+            n = 0,
+            handle = ItHandle,
+            hash = Hash
+        }
+    ].
+
+next_loop(
+    Shard,
+    S = #s{trie = Trie},
+    It = #it{static_index = StaticIdx, ts = TS, compressed_tf = CompressedTF},
+    Iterators,
+    BatchSize,
+    TMax
+) ->
+    TopicStructure =
+        case emqx_ds_lts:reverse_lookup(Trie, StaticIdx) of
+            {ok, Rev} ->
+                Rev;
+            undefined ->
+                throw(#{
+                    msg => "LTS trie missing key",
+                    key => StaticIdx
+                })
+        end,
+    Ctx = #ctx{
+        shard = Shard,
+        s = S,
+        iters = Iterators,
+        topic_structure = TopicStructure,
+        filter = words(CompressedTF),
+        tmax = TMax
+    },
```

Comment:
While this is much better from readability perspective, in my experience it keeping this on the stack (i.e. as arguments) helps significantly in hot loops.

Reply:
This is result of final refactoring. Original version passed everything through function arguments as you described. I didn't find any measurable difference in the throughput or flamegraph before and after introducing context record.

Reply:
I see.

> GC

My point is that this approach usually contributes to higher GC pressure (i.e. garbage needs to be collected much more often).

Reply:
This tuple is never updated (it contains _immutable_ loop context), why would it create any garbage?

Reply:
Ah right, good point, I missed that. Disregard the comment.